File: src/Figure1.py
import numpy as np
import pandas as pd
import matplotlib as mpl

dir_path = '../data/'
data = pd.read_csv(dir_path + "closing_odds.csv")
#data = np.genfromtxt(dir_path + "closing_odds.csv", delimiter=',')

# Fields:
# 1. match_table_id: unique identifier of the game
# 2. league of the game
# 3. match date
# 4. home team
# 5. 90-minute score of home team
# 6. away team
# 7. 90-minute score of away team
# 8. average closing odds home win
# 9. average closing odds draw
# 10. average closing odds away win
# 11. maximum offered closing odds home win
# 12. maximum offered closing odds draw
# 13. maximum offered closing odds away win
# 14. name of bookmaker offering maximum closing odds for home win
# 15. name of bookmaker offering maximum closing odds for draw
# 16. name of bookmaker offering maximum closing odds for away win
# 17. number of available closing odds for home win
# 18. number of available closing odds for draw
# 19. number of available closing odds for away win

n_games = data.shape[0]

[u'match_id', u'league', u'match_date', u'home_team', u'home_score',
       u'away_team', u'away_score', u'avg_odds_home_win', u'avg_odds_draw',
       u'avg_odds_away_win', u'max_odds_home_win', u'max_odds_draw',
       u'max_odds_away_win', u'top_bookie_home_win', u'top_bookie_draw',
       u'top_bookie_away_win', u'n_odds_home_win', u'n_odds_draw',
       u'n_odds_away_win']

leagues = data['league']
n_leagues = pd.unique(data['league']).shape[0]

prior_home = float(sum(data['home_score'] > data['away_score'])) / n_games
prior_draw = float(sum(data['home_score'] == data['away_score'])) / n_games
prior_away = float(sum(data['home_score'] < data['away_score'])) / n_games

print('Total number of games: ' + str(n_games) + "\n")
print('Total number of Leagues:' + str(n_leagues) + "\n")
print('Proportion of Home victories: ' + str(prior_home) + "\n")
print('Proportion of Draws: ' + str(prior_draw) + "\n")
print(('Proportion of Away victories: ' + str(prior_away) + "\n"))

# Calculate accuracy of prediction as a function of the implicit probability
# contained in the odds
odds_bins = np.arange(0, 1, 0.0125)  # probability bins
min_games = 100

# Home victory
p_home = 1 / data['avg_odds_home_win']
p_draw = 1 / data['avg_odds_draw']
p_away = 1 / data['avg_odds_away_win']

home_score = data['home_score']
away_score = data['away_score']

acc_home = []
acc_draw = []
acc_away = []

bin_odds_home_mean = []
bin_odds_draw_mean = []
bin_odds_away_mean = []

for bn in range(0, len(odds_bins) - 2):
    #print("bin " + str(bn + 1) + " from" + str(len(odds_bins) -1) + "\n")
    # Get the data from the bin
    inds_home = np.where((p_home > odds_bins[bn]) & (p_home <= odds_bins[bn + 1]))[0]
    inds_draw = np.where((p_draw > odds_bins[bn]) & (p_draw <= odds_bins[bn + 1]))[0]
    inds_away = np.where((p_away > odds_bins[bn]) & (p_away <= odds_bins[bn + 1]))[0]
    # Get accuracy for home, draw away
    if (len(inds_home) >= min_games):
        acc_home.append(float(sum(home_score[inds_home] > away_score[inds_home])) / len(inds_home))
        bin_odds_home_mean.append(np.mean(p_home[inds_home]))
    if (len(inds_draw) >= min_games):
        acc_draw.append(float(sum(home_score[inds_draw] == away_score[inds_draw])) / len(inds_draw))
        bin_odds_draw_mean.append(np.mean(p_draw[inds_draw]))
    if (len(inds_away) >= min_games):
        acc_away.append(float(sum(home_score[inds_away] < away_score[inds_away])) / len(inds_away))
        bin_odds_away_mean.append(np.mean(p_away[inds_away]))

mpl.pyplot.plot(acc_home, bin_odds_home_mean, '.k')
mpl.pyplot.plot(acc_draw, bin_odds_draw_mean, '.r')
mpl.pyplot.plot(acc_away, bin_odds_away_mean, '.b')
mpl.pyplot.show()

# linear regression
from sklearn import linear_model
from sklearn.metrics import r2_score

home_regr = linear_model.LinearRegression()
draw_regr = linear_model.LinearRegression()
away_regr = linear_model.LinearRegression()

x_home = np.array(bin_odds_home_mean).reshape(-1, 1)
y_home = np.array(acc_home).reshape(-1, 1)

x_draw = np.array(bin_odds_draw_mean).reshape(-1, 1)
y_draw = np.array(acc_draw).reshape(-1, 1)

x_away = np.array(bin_odds_away_mean).reshape(-1, 1)
y_away = np.array(acc_away).reshape(-1, 1)


# fit a linear regression line to the data
home_regr.fit(x_home, y_home)
draw_regr.fit(x_draw, y_draw)
away_regr.fit(x_away, y_away)

home_preds = home_regr.predict(x_home)
draw_preds = draw_regr.predict(x_draw)
away_preds = away_regr.predict(x_away)

print('Home r2: %1.3f, slope: %1.3f, intercept: %1.3f \n' % (r2_score(y_home, home_preds),
      home_regr.coef_[0][0], home_regr.intercept_[0]))

print('Draw r2: %1.3f, slope: %1.3f, intercept: %1.3f \n' % (r2_score(y_draw, draw_preds),
      draw_regr.coef_[0][0], draw_regr.intercept_[0]))

print('Away r2: %1.3f, slope: %1.3f, intercept: %1.3f \n' % (r2_score(y_away, away_preds),
      away_regr.coef_[0][0], away_regr.intercept_[0]))

File: src/Figure1.m
function Figure1()

%% Historical analysis of Bookies prediction accuracy
% Bookmakers prediction power. A historical analysis of 10 years of football
% games shows the tight relationship between the bookmakers’ predictions and
% the actual outcome of football games. 

% comment the next 3 lines if using Matlab
%warning("off")
%pkg load statistics
%pkg load nan

addpath('./aux_files/')
addpath('./aux_plot/')
figs_dir = '../figures/';
dat_dir = '../data/';
file_name = 'closing_odds.csv';

% Fields: 
% 1. match_table_id: unique identifier of the game
% 2. league of the game
% 3. match date
% 4. home team
% 5. 90-minute score of home team
% 6. away team
% 7. 90-minute score of away team
% 8. average closing odds home win
% 9. average closing odds draw
% 10. average closing odds away win
% 11. maximum offered closing odds home win
% 12. maximum offered closing odds draw
% 13. maximum offered closing odds away win
% 14. name of bookmaker offering maximum closing odds for home win
% 15. name of bookmaker offering maximum closing odds for draw
% 16. name of bookmaker offering maximum closing odds for away win
% 17. number of available closing odds for home win
% 18. number of available closing odds for draw
% 19. number of available closing odds for away win
fid = fopen([dat_dir file_name], 'r');
C = textscan(fid, '%s %s %s %s %f %s %f %f %f %f %f %f %f %s %s %s %f %f %f', 'delimiter', ',');
fclose(fid);

dat_hist = [C{5} C{7} C{8} C{9} C{10}];
dat_hist(:, 3:5) = 1 ./ dat_hist(:, 3:5);

leagues = C{2};
nLeagues = length(unique(leagues));
nGames = size(dat_hist, 1);

pHome = sum(dat_hist(:,1) > dat_hist(:,2)) / nGames;
pDraw = sum(dat_hist(:,1) == dat_hist(:,2)) / nGames;
pAway = sum(dat_hist(:,1) < dat_hist(:,2)) / nGames;

fprintf('Total number of games: %d \n', nGames); 
fprintf('Total number of Leagues: %d \n', nLeagues);
fprintf('Proportion of Home victories: %.3f \n', pHome);
fprintf('Proportion of Draws: %.3f \n', pDraw);
fprintf('Proportion of Away victories: %.3f \n', pAway);

%% Calculate accuracy of prediction as a function of the implicit probability 
%% contained in the odds
odds_bins = 0:0.0125:1; % probability bins
nGames2Bet = 100;

% Home victory
val_bin = 1; % counter for valid bins
for bn = 1 : length(odds_bins) - 1
   
    % Get the data from the bin
    dat_bin = dat_hist(dat_hist(:, 3) > odds_bins(bn) & dat_hist(:, 3) <= odds_bins(bn + 1), :);
    % Get accuracy for home, draw away
    if isempty(dat_bin) || size(dat_bin,1) < nGames2Bet
        continue
    end
        
    % Home wins
    acc_home(val_bin) = sum(dat_bin(:,1) > dat_bin(:,2)) / size(dat_bin,1); %#ok
    bin_odds_home_mean(val_bin) = nanmean(dat_bin(:,3)); %#ok
    val_bin = val_bin + 1; 
end

% Draw
val_bin = 1;
for bn = 1 : length(odds_bins) - 1
   
    % Get the data from the bin
    dat_bin = dat_hist(dat_hist(:, 4) > odds_bins(bn) & dat_hist(:, 4) <= odds_bins(bn + 1), :);
    % Get accuracy for home, draw away
    if isempty(dat_bin) || size(dat_bin,1) < nGames2Bet
        continue
    end

    nGms(val_bin) =  size(dat_bin,1); %#ok
    acc_draws(val_bin) = sum(dat_bin(:,1) == dat_bin(:,2)) / size(dat_bin,1);  %#ok
    bin_odds_draws_mean(val_bin) = nanmean(dat_bin(:,4)); %#ok
    val_bin = val_bin + 1;
end

% Away Victory
val_bin = 1;
for bn = 1 : length(odds_bins) - 1
   
    % Get the data from the bin
    dat_bin = dat_hist(dat_hist(:, 5) > odds_bins(bn) & dat_hist(:, 5) <= odds_bins(bn + 1), :);
    % Get accuracy for home, draw away
    if isempty(dat_bin) || size(dat_bin,1) < nGames2Bet
        continue
    end
    
    % Away wins
    acc_away(val_bin) = sum(dat_bin(:,1) < dat_bin(:,2)) / size(dat_bin,1); %#ok
    bin_odds_away_mean(val_bin) = nanmean(dat_bin(:,5)); %#ok
    val_bin = val_bin + 1; 
end

%% Plot results
f1 = figure(1); clf;
set(gcf, 'Position', [0 0 1200 800], 'InvertHardCopy', 'on', 'PaperPositionMode', 'auto');
hold on;

plot(bin_odds_home_mean, acc_home, 'ok', 'MarkerSize', 5, 'MarkerFace', 'k')
plot(bin_odds_draws_mean, acc_draws, 'om', 'MarkerSize', 5, 'MarkerFace', 'm')
plot(bin_odds_away_mean, acc_away, 'ob', 'MarkerSize', 5, 'MarkerFace', 'b')

xlabel('Estimated probability [ 1 / odds]')
ylabel('p(Correct)')

set(gca, 'FontSize', 20)

legend('Home Victory', 'Draw', 'Away Victory', 'Location', 'SouthEast'); legend boxoff;

print(f1, '-dpng', [figs_dir 'Figure1.png'])
print(f1, '-depsc', [figs_dir 'Figure1.eps'])

%% Linear regression

[slp, itrc, rsq, ~, ~] = run_corr (bin_odds_home_mean, acc_home);
fprintf('Home r2: %1.3f, slope: %1.3f, intercept: %1.3f \n', rsq, slp, itrc); 

[slp, itrc, rsq, ~, ~] = run_corr (bin_odds_draws_mean, acc_draws);
fprintf('Draw r2: %1.3f, slope: %1.3f, intercept: %1.3f \n', rsq, slp, itrc); 

[slp, itrc, rsq, ~, ~] = run_corr (bin_odds_away_mean, acc_away);
fprintf('Away r2: %1.3f, slope: %1.3f, intercept: %1.3f \n', rsq, slp, itrc); 


